include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort_one_wg.h

```diff
@@ -249,11 +235,41 @@ struct __subgroup_radix_sort
                                 {
                                     const uint16_t __r = __indices[__i];
                                     if (__r < __n)
-                                        __src[__r] = __values[__i];
+                                        __src[__r] = ::std::move(__values.__v[__i]);
                                 }
+                                //destroy values
+                                _ONEDPL_PRAGMA_UNROLL
+                                for (uint16_t __i = 0; __i < __block_size; ++__i)
+                                    __values.__v[__i].~_ValT();
+
+                                //destroy values in exchange buffer
+                                _ONEDPL_PRAGMA_UNROLL
+                                for (uint16_t __i = 0; __i < __block_size; ++__i)
+                                    __exchange_lacc[__wi * __block_size + __i].~_ValT();
+
                                 return;
                             }
-                            __to_blocked(__it, __wi, __exchange_lacc, __values, __indices);
+
+                            //3.1 data exchange
+                            if(__begin_bit == __radix) //the first sort iteration
+                            {
+                                _ONEDPL_PRAGMA_UNROLL
+                                for (uint16_t __i = 0; __i < __block_size; ++__i)
+                                    new(&__exchange_lacc[__indices[__i]]) _ValT(::std::move(__values.__v[__i]));
```

Comment:
Can we be moving data we have never constructed?
(again, when going past `__n`)

Similar questions below.

Reply:
yes, it seems we have to check "< n"...

Reply:
yes, it seems we have to check "< n"

Reply:
fixed